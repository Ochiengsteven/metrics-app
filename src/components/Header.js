import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
// eslint-disable-next-line import/no-extraneous-dependencies
import debounce from 'lodash.debounce';
import './assets/css/header.css';

export default function Header({ className, onSearch }) {
  const [searchQuery, setSearchQuery] = useState('');

  // Create a debounced search function to delay the filtering
  const debouncedSearch = debounce((query) => {
    onSearch(query);
  }, 300);

  const handleInputChange = (event) => {
    const query = event.target.value;
    setSearchQuery(query);
    debouncedSearch(query);
  };

  useEffect(() => {
    if (searchQuery === '') {
      onSearch(searchQuery);
    }
  }, [searchQuery, onSearch]);

  return (
    <div className={className}>
      <h1 className="text-5xl font-extrabold w-full pb-4">CryptoView</h1>
      <h3 className="text-3xl font-semibold w-full back3">
        Research the&nbsp;
        <br />
        crypto
        <br />
        market
      </h3>
      <input
        type="text"
        placeholder="Search by pair..."
        className="border rounded p-2 text-black focus:outline-none mt-2 w-5/6"
        value={searchQuery}
        onChange={handleInputChange}
      />
    </div>
  );
}

// Prop validation
Header.propTypes = {
  className: PropTypes.string.isRequired,
  onSearch: PropTypes.func.isRequired,
};
