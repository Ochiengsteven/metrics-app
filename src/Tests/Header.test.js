import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Header from '../components/Header';

// Mock the debounce function
jest.mock('lodash.debounce', () => jest.fn((fn) => fn));

const mockStore = configureStore([]);

describe('Header Component', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      assets: {
        status: 'succeeded',
        error: null,
      },
    });
  });

  it('renders loading state', () => {
    store = mockStore({
      assets: {
        status: 'loading',
        error: null,
      },
    });

    const { container } = render(
      <Provider store={store}>
        <Header className="test-class" onSearch={() => {}} />
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders error state', () => {
    store = mockStore({
      assets: {
        status: 'failed',
        error: 'Error message',
      },
    });

    const { container } = render(
      <Provider store={store}>
        <Header className="test-class" onSearch={() => {}} />
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });
});
