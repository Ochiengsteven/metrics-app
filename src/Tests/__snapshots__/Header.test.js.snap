// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component renders correctly 1`] = `
<div>
  <div
    class="test-class"
  >
    <h1
      class="text-5xl font-extrabold w-full pb-4"
    >
      CryptoView
    </h1>
    <h3
      class="text-3xl font-semibold w-full back3"
    >
      Research the 
      <br />
      crypto
      <br />
      market
    </h3>
    <input
      class="border rounded p-2 text-black focus:outline-none mt-2 w-5/6"
      placeholder="Search by pair..."
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`Header Component renders error state 1`] = `
<div>
  <div
    class="test-class"
  >
    <h1
      class="text-5xl font-extrabold w-full pb-4"
    >
      CryptoView
    </h1>
    <h3
      class="text-3xl font-semibold w-full back3"
    >
      Research the 
      <br />
      crypto
      <br />
      market
    </h3>
    <input
      class="border rounded p-2 text-black focus:outline-none mt-2 w-5/6"
      placeholder="Search by pair..."
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`Header Component renders loading state 1`] = `
<div>
  <div
    class="test-class"
  >
    <h1
      class="text-5xl font-extrabold w-full pb-4"
    >
      CryptoView
    </h1>
    <h3
      class="text-3xl font-semibold w-full back3"
    >
      Research the 
      <br />
      crypto
      <br />
      market
    </h3>
    <input
      class="border rounded p-2 text-black focus:outline-none mt-2 w-5/6"
      placeholder="Search by pair..."
      type="text"
      value=""
    />
  </div>
</div>
`;
